File: conformance/code/individual/src/main/java/com/digit/individual/repository/ConfigRepository.java
package com.digit.individual.repository;

import com.digit.individual.model.Config;
import org.digit.tracer.observability.ObservabilityMetrics;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.jdbc.support.GeneratedKeyHolder;
import org.springframework.jdbc.support.KeyHolder;
import org.springframework.stereotype.Repository;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.List;

/**
 * JDBC repository for per-tenant validation config (table individual_config_v3). uniquenesscriteria
 * is a raw jsonb column read/written as JSON text.
 */
@Repository
public class ConfigRepository {

    private static final String T = "individual_config_v3";

    private final JdbcTemplate jdbc;
    private final ObservabilityMetrics metrics;

    public ConfigRepository(JdbcTemplate jdbc, ObservabilityMetrics metrics) {
        this.jdbc = jdbc;
        this.metrics = metrics;
    }

    private final RowMapper<Config> mapper = (RowMapper<Config>) (ResultSet rs, int n) -> {
        Config c = new Config();
        c.setId(rs.getLong("id"));
        c.setTenantId(rs.getString("tenantid"));
        c.setMobileRegex(rs.getString("mobileregex"));
        c.setNameRegex(rs.getString("nameregex"));
        c.setUniquenessCriteria(rs.getString("uniquenesscriteria"));
        c.setVersion(rs.getInt("version"));
        c.setCreatedBy(rs.getString("createdBy"));
        c.setModifiedBy(rs.getString("modifiedBy"));
        c.setCreatedTime(rs.getLong("createdTime"));
        c.setModifiedTime(rs.getLong("modifiedTime"));
        c.setRequestId(rs.getString("requestid"));
        return c;
    };

    private static final String COLS = "id, tenantid, mobileregex, nameregex, uniquenesscriteria, version, "
            + "\"createdBy\", \"modifiedBy\", \"createdTime\", \"modifiedTime\", requestid";

    public Config getByTenant(String tenantId) {
        boolean ok = true;
        try {
            List<Config> rows = jdbc.query("SELECT " + COLS + " FROM " + T + " WHERE tenantid = ? LIMIT 1",
                    mapper, tenantId);
            return rows.isEmpty() ? null : rows.get(0);
        } catch (RuntimeException e) {
            ok = false;
            throw e;
        } finally {
            metrics.recordDbOperation("SELECT", T, ok);
        }
    }

    /**
     * Like {@link #getByTenant} but takes a row lock (SELECT ... FOR UPDATE) so a concurrent upsert
     * for the same tenant serialises behind it. Used only by the upsert read-modify-write to keep the
     * version counter race-free. Must run inside the request transaction (opened by
     * TenantTransactionFilter) so the lock is held across the subsequent update.
     */
    public Config getByTenantForUpdate(String tenantId) {
        boolean ok = true;
        try {
            List<Config> rows = jdbc.query("SELECT " + COLS + " FROM " + T + " WHERE tenantid = ? LIMIT 1 FOR UPDATE",
                    mapper, tenantId);
            return rows.isEmpty() ? null : rows.get(0);
        } catch (RuntimeException e) {
            ok = false;
            throw e;
        } finally {
            metrics.recordDbOperation("SELECT", T, ok);
        }
    }

    /** Inserts a new config row; populates the generated id back onto cfg. */
    public void insert(Config cfg) {
        boolean ok = true;
        try {
            KeyHolder kh = new GeneratedKeyHolder();
            jdbc.update(con -> {
                PreparedStatement ps = con.prepareStatement(
                        "INSERT INTO " + T + " (tenantid, mobileregex, nameregex, uniquenesscriteria, version, "
                                + "\"createdBy\", \"modifiedBy\", \"createdTime\", \"modifiedTime\", requestid) "
                                + "VALUES (?,?,?,?::jsonb,?,?,?,?,?,?)",
                        new String[]{"id"});
                ps.setString(1, cfg.getTenantId());
                ps.setString(2, cfg.getMobileRegex());
                ps.setString(3, cfg.getNameRegex());
                ps.setString(4, cfg.getUniquenessCriteria());
                ps.setInt(5, cfg.getVersion());
                ps.setString(6, cfg.getCreatedBy());
                ps.setString(7, cfg.getModifiedBy());
                ps.setLong(8, cfg.getCreatedTime());
                ps.setLong(9, cfg.getModifiedTime());
                ps.setString(10, cfg.getRequestId());
                return ps;
            }, kh);
            Number key = kh.getKey();
            if (key != null) {
                cfg.setId(key.longValue());
            }
        } catch (RuntimeException e) {
            ok = false;
            throw e;
        } finally {
            metrics.recordDbOperation("INSERT", T, ok);
        }
    }

    /** Overwrites the row identified by cfg.id. Callers preserve immutable audit fields. */
    public void update(Config cfg) {
        boolean ok = true;
        try {
            jdbc.update("UPDATE " + T + " SET tenantid=?, mobileregex=?, nameregex=?, uniquenesscriteria=?::jsonb, "
                            + "version=?, \"createdBy\"=?, \"modifiedBy\"=?, \"createdTime\"=?, \"modifiedTime\"=?, "
                            + "requestid=? WHERE id=?",
                    cfg.getTenantId(), cfg.getMobileRegex(), cfg.getNameRegex(), cfg.getUniquenessCriteria(),
                    cfg.getVersion(), cfg.getCreatedBy(), cfg.getModifiedBy(), cfg.getCreatedTime(),
                    cfg.getModifiedTime(), cfg.getRequestId(), cfg.getId());
        } catch (RuntimeException e) {
            ok = false;
            throw e;
        } finally {
            metrics.recordDbOperation("UPDATE", T, ok);
        }
    }
}

File: conformance/code/individual-go/internal/repository/config_repository.go
package repository

import (
	"context"
	"time"

	"individual/internal/models"

	tenantdb "github.com/digitnxt/digit3/src/libraries/tenant-migration/tenantdb"
	tracerobs "github.com/digitnxt/digit3/src/libraries/tracer/observability"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"gorm.io/gorm"
	"gorm.io/gorm/clause"
)

type ConfigRepository interface {
	GetByTenant(ctx context.Context, tenantID string) (*models.Config, error)
	GetByTenantForUpdate(ctx context.Context, tenantID string) (*models.Config, error)
	Insert(ctx context.Context, cfg *models.Config) error
	Update(ctx context.Context, cfg *models.Config) error
}

type configRepository struct {
	db *gorm.DB
}

func NewConfigRepository(db *gorm.DB) ConfigRepository {
	return &configRepository{db: db}
}

func (r *configRepository) GetByTenant(ctx context.Context, tenantID string) (*models.Config, error) {
	tracer := otel.Tracer("individual-repository")
	ctx, span := tracer.Start(ctx, "db.config.getByTenant")
	defer span.End()
	span.SetAttributes(
		attribute.String("db.table", "individual_config_v3"),
		attribute.String("tenant.id", tenantID),
	)
	start := time.Now()

	var cfg models.Config
	db := tenantdb.GetTenantDB(ctx, r.db)
	err := db.WithContext(ctx).Where("tenantid = ?", tenantID).First(&cfg).Error

	duration := time.Since(start)
	tracerobs.RecordDBOperation(ctx, "SELECT", "individual_config_v3", duration, err == nil || err == gorm.ErrRecordNotFound)
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			span.SetStatus(codes.Ok, "")
			return nil, nil
		}
		span.RecordError(err)
		span.SetStatus(codes.Error, "failed to get config by tenant")
		return nil, err
	}
	span.SetStatus(codes.Ok, "")
	return &cfg, nil
}

// GetByTenantForUpdate is GetByTenant plus a row lock (SELECT ... FOR UPDATE) so
// a concurrent upsert for the same tenant serialises behind it. Used only by
// Upsert to make the read-modify-write (version bump) race-free; plain reads use
// GetByTenant. Relies on the per-request transaction (tenantdb middleware) so the
// lock is held until the request commits.
func (r *configRepository) GetByTenantForUpdate(ctx context.Context, tenantID string) (*models.Config, error) {
	tracer := otel.Tracer("individual-repository")
	ctx, span := tracer.Start(ctx, "db.config.getByTenantForUpdate")
	defer span.End()
	span.SetAttributes(attribute.String("tenant.id", tenantID))
	start := time.Now()

	var cfg models.Config
	db := tenantdb.GetTenantDB(ctx, r.db)
	err := db.WithContext(ctx).
		Clauses(clause.Locking{Strength: "UPDATE"}).
		Where("tenantid = ?", tenantID).
		First(&cfg).Error

	tracerobs.RecordDBOperation(ctx, "SELECT", "individual_config_v3", time.Since(start), err == nil || err == gorm.ErrRecordNotFound)
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			span.SetStatus(codes.Ok, "")
			return nil, nil
		}
		span.RecordError(err)
		span.SetStatus(codes.Error, "failed to lock config by tenant")
		return nil, err
	}
	span.SetStatus(codes.Ok, "")
	return &cfg, nil
}

// Insert creates a new config row. cfg.ID is populated by GORM after the call.
func (r *configRepository) Insert(ctx context.Context, cfg *models.Config) error {
	tracer := otel.Tracer("individual-repository")
	ctx, span := tracer.Start(ctx, "db.config.insert")
	defer span.End()
	span.SetAttributes(attribute.String("db.table", "individual_config_v3"))
	start := time.Now()

	db := tenantdb.GetTenantDB(ctx, r.db)
	err := db.WithContext(ctx).Create(cfg).Error

	duration := time.Since(start)
	tracerobs.RecordDBOperation(ctx, "INSERT", "individual_config_v3", duration, err == nil)
	if err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, "failed to insert config")
		return err
	}
	span.SetStatus(codes.Ok, "")
	return nil
}

// Update overwrites the row identified by cfg.ID with the values in cfg.
// Callers are responsible for preserving immutable fields (CreatedBy/CreatedTime)
// before calling.
func (r *configRepository) Update(ctx context.Context, cfg *models.Config) error {
	tracer := otel.Tracer("individual-repository")
	ctx, span := tracer.Start(ctx, "db.config.update")
	defer span.End()
	span.SetAttributes(attribute.String("db.table", "individual_config_v3"))
	start := time.Now()

	db := tenantdb.GetTenantDB(ctx, r.db)
	err := db.WithContext(ctx).Save(cfg).Error

	duration := time.Since(start)
	tracerobs.RecordDBOperation(ctx, "UPDATE", "individual_config_v3", duration, err == nil)
	if err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, "failed to update config")
		return err
	}
	span.SetStatus(codes.Ok, "")
	return nil
}
